### PR TITLE
 Add small pr guidelines

### LINKED ISSUE #2812: TODO: Write up small PRs recommendations DEVELOPMENT.md

### PR DESCRIPTION
**Pull request by tejal29**
Fixes #2812 

**Description**
I this PR, add guidelines for keeping PRs small, 
1. for config changes
2.  Big feature requests.

**User facing changes**
1. DEVELOPMENT.md links to writing small prs. 

**Next PRs.**

- [ ] How to keep your PRs small when Addressing bug fixes.
      Explain how one should address only one bug in a PR. 

**Submitter Checklist**

These are the criteria that every PR should meet, please check them off as you
review them:

- [ n/a] Includes [unit tests](../DEVELOPMENT.md#creating-a-pr)
- [n/a ] Mentions any output changes.
- [ X ] Adds documentation as needed: user docs, YAML reference, CLI reference.
- [  n/a ] Adds integration tests if needed.

_See [the contribution guide](../CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a example to the `examples` dir](https://github.com/GoogleContainerTools/skaffold/tree/master/examples), please copy them to [`integration/examples`](https://github.com/GoogleContainerTools/skaffold/tree/master/integration/examples)
- Every new example added in [`integration/examples` dir](https://github.com/GoogleContainerTools/skaffold/tree/master/integration/examples), should be tested in [integration test](https://github.com/GoogleContainerTools/skaffold/tree/master/integration)

**Reviewer Notes**

- [ ] The code flow looks good. 
- [ ] Unit test added.
- [ ] User facing changes look good.

**Release Notes**

Describe any user facing changes here so maintainer can include it in the release notes, or delete this block.

```
Examples of user facing changes:
- Skaffold config changes like
  e.g. "Add buildArgs to `Kustomize` deployer skaffold config."
- Bug fixes
  e.g. "Improve skaffold init behaviour when tags are used in manifests"
- Any changes in skaffold behavior
  e.g. "Artiface cachine is turned on by default."

```

### CODE REVIEW

File: docs/community/small-prs.md [hunk 1]
```diff
@@ -0,0 +1,75 @@
+# Pull Request Guidelines
+
+Writing good pull requests increase development velocity and minimize frustration.
```
Comment (dgageot) [inline docs/community/small-prs.md]: increases and minimizes?

File: docs/community/small-prs.md [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ 
+If pull requests are small they will result in being opened and merged quickly.
```
Comment (balopat) [inline docs/community/small-prs.md]:
```suggestion
Small pull request will be merged quicker.
```

File: docs/community/small-prs.md [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+Having big pull requests that bounce back and forth between developers
+and reviewers can slow progress down significantly, causing developers to waste a 
+lot of time dealing with merge conflicts.
+
+Big pull requests also increase the risk of breaking things; since changes are so big, 
+ - it’s hard to scrutinize all the changes and
```
Comment (balopat) [inline docs/community/small-prs.md]:
```suggestion
 - it’s hard to truly understand all the changes and 
```

File: docs/community/small-prs.md [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+Big pull requests also increase the risk of breaking things; since changes are so big, 
+ - it’s hard to scrutinize all the changes and 
+ - test them for regressions.
+
+In order to promote good PR quality, the skaffold team will push back
+against big PRs. See [0](https://github.com/GoogleContainerTools/skaffold/pull/2750#pullrequestreview-283621241), [1](https://github.com/GoogleContainerTools/skaffold/pull/2917#pullrequestreview-291415741)
+
+## Breaking down Pull Requests
+
+### Feature development.
```
Comment (balopat) [inline docs/community/small-prs.md]:
```suggestion
### Feature development
```

File: docs/community/small-prs.md [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+Adding a new command or skaffold config results in large number of changes due to
+all the boilerplate code required for adding a new command, generating docs etc.
+
+It makes sense to break down these changes into 2 categories:
+1. Incremental code changes which are not accessed by control flow.
```
Comment (balopat) [inline docs/community/small-prs.md]:
```suggestion
1. "Invisible changes": Incremental, small code changes 
     1. refactoring: either keep the functionality the same but refactoring the design or
     1. partial implementations: unit tested, new behavior, that is not yet accessible by the main control flow.
```

File: docs/community/small-prs.md [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+2. User affecting changes like
```
Comment (balopat) [inline docs/community/small-prs.md]:
```suggestion
2. "Visible changes": User affecting changes, for example
```

File: docs/community/small-prs.md [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+   1. Adding a new command which exercises the above code changes
+   1. Adding config change and generating docs.
+
+#### Config Change Example   
+Below is an example of how you can break a config change into small individual PRs.
+
+A contributor wanted to support [custom build args to `kustomize` deployer](https://github.com/GoogleContainerTools/skaffold/issues/2488).
+This required, skaffold to support a `buildArgs` field to `deploy.kustomize` config. Plumb the config to `kustomize` command.
```
Comment (balopat) [inline docs/community/small-prs.md]:
```suggestion
This required skaffold to add the `deploy.kustomize.buildArgs` field and then plumb the arguments to the `kustomize` command.
```

File: docs/community/small-prs.md [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+   
+   Note: The code added in this PR does not get exercised other than in tests.
+2. Then add render implementation for each [kubectl deployer](https://github.com/GoogleContainerTools/skaffold/pull/2943), kustomize and helm.
+3. Add a [render command](https://github.com/GoogleContainerTools/skaffold/pull/2942)
+
+Remember, its ok to land code which is not exercised. However please mention the follow up work
+in **Next PRs** section when creating a Pull Request [e.g. here.](https://github.com/GoogleContainerTools/skaffold/pull/2811)
+
+
+Finally, please use our best judgement when submitting pull requests.
```
Comment (balopat) [inline docs/community/small-prs.md]:
```suggestion
Finally, please use your best judgement when submitting pull requests, these rules might not always work for you - we would love to hear that! 
```

**Comment by tejal29** [comment]
> Implementing the full functionality sometimes makes a lot of sense, so that you can get feedback regarding the design while implementing it

We would want to encourage sending a `design proposal` for feedback on design or large changes. Having a design proposal does help to think about breaking down in smaller PRs since we have a section for it.
I thought i personally spent a lot of time rebasing my master PR i had for #2750 and just abandoned it.  Nick also did the same thing with #2911

With that said, i do understand scenarios like this #2978 
Do you think #2978, #2917, #2911 all could be an example of having a proposal?

I feel it might send conflicting messages but open for discussion.

**Pull request merged**